catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/QueryOperations.java

```diff
@@ -324,10 +247,66 @@ public QueryResponse replaceSourceId(QueryResponse queryResponse) {
         }
         newResponse.setHits(queryResponse.getHits());
         newResponse.closeResultQueue();
-        LOGGER.debug("EXITING: replaceSourceId()");
+        LOGGER.trace("EXITING: replaceSourceId()");
         return newResponse;
     }
 
+    boolean canAccessSource(FederatedSource source, QueryRequest request) {
+        Map<String, Set<String>> securityAttributes = source.getSecurityAttributes();
+        if (securityAttributes.isEmpty()) {
+            return true;
+        }
+
+        Object requestSubject = request.getProperties()
+                .get(SecurityConstants.SECURITY_SUBJECT);
+        if (requestSubject instanceof ddf.security.Subject) {
+            Subject subject = (Subject) requestSubject;
+
+            KeyValueCollectionPermission kvCollection = new KeyValueCollectionPermission(
+                    CollectionPermission.READ_ACTION,
+                    securityAttributes);
+            boolean isPermitted = subject.isPermitted(kvCollection);
+            if (isPermitted) {
+                SecurityLogger.audit("Subject is permitted to access source {}", source.getId());
+            } else {
+                SecurityLogger.audit("Subject is not permitted to access source {}",
+                        source.getId());
+            }
+            return isPermitted;
+        }
+        return false;
+    }
+
+    /**
+     * Determines if the local catlog provider's source ID is included in the list of source IDs. A
+     * source ID in the list of null or an empty string are treated the same as the local source's
+     * actual ID being in the list.
+     *
+     * @param sourceIds the list of source IDs to examine
+     * @return true if the list includes the local source's ID, false otherwise
+     */
+    boolean includesLocalSources(Set<String> sourceIds) {
```

Comment:
Not 100% sure, but it looks like this method is only used by `QuerySources`. Could we move it to that class?